kernel/src/drivers/ps2/mod.rs

```diff
@@ -3,82 +3,85 @@ pub mod io;
 use drivers::ps2::io::*;
 use spin::Mutex;
 
-pub const DEVICE_ENABLED_FLAG: u8 = 1 << 0;
-pub const DEVICE_SECOND_FLAG: u8 = 1 << 1;
-pub const DEVICE_AVAILABLE_FLAG: u8 = 1 << 2;
-
 pub const RESEND: u8 = 0xFE;
 pub const ACK: u8 = 0xFA;
-pub const SELF_TEST_PASSED: u8 = 0xAA;
 
-pub static PS2: Mutex<Ps2Controller> = Mutex::new(Ps2Controller::new());
+lazy_static! {
+    pub static ref PS2: Mutex<Ps2Controller> = Mutex::new(Ps2Controller::new());
+}
+
+bitflags! {
+    pub struct Ps2ConfigFlags: u8 {
+        const PORT_INTERRUPT_1 = 1 << 0;
+        const PORT_INTERRUPT_2 = 1 << 1;
+        const SYSTEM = 1 << 2;
+        const KEYBOARD_LOCK = 1 << 4;
+        const PORT_OUTPUT_FULL_2 = 1 << 5;
+        const PORT_TRANSLATION_1 = 1 << 6;
+    }
+}
 
 /// Represents the PS2 master controller
 pub struct Ps2Controller {
     pub initialized: bool,
     pub devices: [Ps2Device; 2],
-    config: Ps2Config,
+    config: Ps2ConfigFlags,
 }
 
 impl Ps2Controller {
-    const fn new() -> Self {
+    fn new() -> Self {
         Ps2Controller {
             initialized: false,
             devices: [
-                Ps2Device {
-                    flags: 0,
-                },
-                Ps2Device {
-                    flags: DEVICE_SECOND_FLAG,
-                }],
-            config: Ps2Config {
-                data: 0,
-            },
+                Ps2Device::new(DevicePort::Keyboard),
+                Ps2Device::new(DevicePort::Mouse),
+            ],
+            config: Ps2ConfigFlags::empty(),
         }
     }
 
     /// Initializes this PS2 controller
-    pub fn initialize(&mut self) {
+    pub fn initialize(&mut self) -> Result<(), Ps2Error> {
```

Comment:
Maybe we can check if it has previously been initialized? Although, personally I'm sort of iffy about having an initialized field anyway...